README.md
[![Build Status](https://travis-ci.org/rvw-org/rvw.svg?branch=master)](https://travis-ci.org/rvw-org/rvw)

## rvw

Development of **rvw** package started as R Vowpal Wabbit (Google Summer of Code 2018) [project](https://summerofcode.withgoogle.com/projects/#5511455416254464).

**Vowpal Wabbit** is an online machine learning system that is known for its speed and scalability and is widely used in research and industry.

This package aims to bring its functionality to **R**.

## Installation

### From Source 

First you have to install **Vowpal Wabbit** from [here](https://github.com/JohnLangford/vowpal_wabbit#getting-the-code).

Next, once the required library isinstalled, you can install the **rvw** package using `remotes`:

```r
install.packages("remotes")  ## or devtools
remotes::install_github("rvw-org/rvw")
```

### Using Docker

We use [Docker](https://www.docker.com) for the [Travis CI](https://www.travis-ci.org) tests, and also provide a container
for deployment. Do 

```sh
docker pull rvowpalwabbit/run                 ## one time 
docker run --rm -ti rvowpalwabbit/run bash    ## launch container
```

to start the container with `rvw` installed.  See the 
[Boettiger and Eddelbuettel RJournal paper](https://journal.r-project.org/archive/2017/RJ-2017-065/index.html)
for more on Docker for R, and the [Rocker Project](https://www.rocker-project.org) used here.

## Example 

In this example we will try to predict age groups (based on number of abalone shell rings) from physical measurements. We will use Abalone Data Set from [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/datasets/Abalone).

First we prepare our data:

```r
library(mltools)
library(rvw)

set.seed(1)
aburl = 'http://archive.ics.uci.edu/ml/machine-learning-databases/abalone/abalone.data'
abnames = c('sex','length','diameter','height','weight.w','weight.s','weight.v','weight.sh','rings')
abalone = read.table(aburl, header = F , sep = ',', col.names = abnames)
data_full <- abalone

# Split number of rings into groups with equal (as possible) number of observations
data_full$group <- bin_data(data_full$rings, bins=3, binType = "quantile")
group_lvls <- levels(data_full$group)
levels(data_full$group) <- c(1, 2, 3)

# Prepare indices to split data
ind_train <- sample(1:nrow(data_full), 0.8*nrow(data_full))
# Split data into train and test subsets
df_train <- data_full[ind_train,]
df_test <- data_full[-ind_train,]
```

Then we set up a *Vowpal Wabbit* model:
```r
vwmodel <- vwsetup(option = "ect", num_classes = 3)
```

* *option = "ect"* - we will use [Error Correcting Tournament](https://github.com/JohnLangford/vowpal_wabbit/wiki/Error-Correcting-Tournament-(ect)-multi-class-example) algorithm to train multiclass classification model;
* *num_classes = 3* - number of classes in our data;

Now we start training:

```r
vwtrain(vwmodel, data = df_train,
        namespaces = list(NS1 = list("sex", "rings"),
                          NS2 = list("weight.w","weight.s","weight.v","weight.sh", "diameter", "length", "height")),
        targets = "group"
)
```
And we get: `average loss = 0.278060`

* *namespaces* - We will split our features into two namespaces `NS1` and `NS2`;
* *targets = "group"* - ground truth labels;


And finally compute predictions using trained model:

```r
predict.vw(vwmodel, data = df_test)
```
Here we get: `average loss = 0.221292`

We can add more learning algorithms to our model. For example we want to use *boosting* algorithm with 100 "weak" learners. Then we will just add this option to our model and train again:

```r
vwmodel <- add_option(vwmodel, option = "boosting", num_learners=100)

vwtrain(vwmodel, data = df_train,
        namespaces = list(NS1 = list("sex", "rings"),
                          NS2 = list("weight.w","weight.s","weight.v","weight.sh", "diameter", "length", "height")),
        targets = "group"
)
```
We get: `average loss = 0.229273`

And compute predictions:

```r
predict.vw(vwmodel, data = df_test)
```
Finally we get: `average loss = 0.081340`

In order to inspect parameters of our model we can simply print it:

```r
vwmodel
```

```
	Vowpal Wabbit model
Learning algorithm:   sgd 
Working directory:   /var/folders/yx/6949djdd3yb4qsw7x_95wfjr0000gn/T//RtmpjO3DD1 
Model file:   /var/folders/yx/6949djdd3yb4qsw7x_95wfjr0000gn/T//RtmpjO3DD1/vw_1534253637_mdl.vw 
General parameters: 
	 random_seed :   0 
	 ring_size :  Not defined
	 holdout_off :   FALSE 
	 holdout_period :   10 
	 holdout_after :   0 
	 early_terminate :   3 
	 loss_function :   squared 
	 link :   identity 
	 quantile_tau :   0.5 
Feature parameters: 
	 bit_precision :   18 
	 quadratic :  Not defined
	 cubic :  Not defined
	 interactions :  Not defined
	 permutations :   FALSE 
	 leave_duplicate_interactions :   FALSE 
	 noconstant :   FALSE 
	 feature_limit :  Not defined
	 ngram :  Not defined
	 skips :  Not defined
	 hash :  Not defined
	 affix :  Not defined
	 spelling :  Not defined
Learning algorithms / Reductions: 
	 ect :
		 num_classes :   3 
	 boosting :
		 num_learners :   100 
		 gamma :   0.1 
		 alg :   BBM 
Optimization parameters: 
	 adaptive :   TRUE 
	 normalized :   TRUE 
	 invariant :   TRUE 
	 adax :   FALSE 
	 sparse_l2 :   0 
	 l1_state :   0 
	 l2_state :   1 
	 learning_rate :   0.5 
	 initial_pass_length :  Not defined
	 l1 :   0 
	 l2 :   0 
	 no_bias_regularization :  Not defined
	 feature_mask :  Not defined
	 decay_learning_rate :   1 
	 initial_t :   0 
	 power_t :   0.5 
	 initial_weight :   0 
	 random_weights :  Not defined
	 normal_weights :  Not defined
	 truncated_normal_weights :  Not defined
	 sparse_weights :   FALSE 
	 input_feature_regularizer :  Not defined
Model evaluation. Training: 
	 num_examples :   3341 
	 weighted_example_sum :   3341 
	 weighted_label_sum :   0 
	 avg_loss :   0.2292727 
	 total_feature :   33408 
Model evaluation. Testing: 
	 num_examples :   836 
	 weighted_example_sum :   836 
	 weighted_label_sum :   0 
	 avg_loss :   0.08133971 
	 total_feature :   8360 
```


ChangeLog
2018-10-13  Dirk Eddelbuettel  <edd@debian.org>

	* docker/ci/Dockerfile: Install mltools (and r-cran-matrix)

2018-09-12  Dirk Eddelbuettel  <edd@debian.org>

	* README.md: Added brief Docker documentation

	* .travis.yml: Update container name

2018-09-11  Dirk Eddelbuettel  <edd@debian.org>

	* docker/run/Dockerfile: Added for deployment

2018-09-10  Dirk Eddelbuettel  <edd@debian.org>

	* .travis.yml: Updated container reference

2018-09-09  Dirk Eddelbuettel  <edd@debian.org>

	* docker/ci/Dockerfile: Add Dockerfile for Travis
	* .travis.yml: Use Docker container in tests

---- a lot omitted here, history from rvw-legacy below ---

2017-03-19  Dirk Eddelbuettel  <edd@debian.org>

	* DESCRIPTION (Version, Date): Roll minor version

	* R/dt2vw.R (dt2vw): Make a copy of the dataset;
	test for NA values in input data set

2017-03-12  Selim Raboudi  <selim.raboudi@axa.com>

	* R/dt2vw.R (dt2vw): Allow for escaped variable names

2016-09-02  Dirk Eddelbuettel  <edd@debian.org>

	* demo/vw_example_4.R: Added Rborist

2016-08-30  Dirk Eddelbuettel  <edd@debian.org>

	* demo/vw_example_4.R: Added ranger

2016-08-29  Dirk Eddelbuettel  <edd@debian.org>

	* README.md: Added

2016-08-23  Dirk Eddelbuettel  <edd@debian.org>

	* demo/vw_example_4.R: Rewritten/extended, now with ctree and gbm

	* demo/vw_example_5.R: Plot predicted vs actual for regression example

	* R/vw.R (vw): Set a AUC fallback value

2016-08-22  Dirk Eddelbuettel  <edd@debian.org>

	* DESCRIPTION (License): Use dual license with GPL (>= 2) for my code,
	and BSD 3-clause for existing code

	* NAMESPACE: All importing now via importFrom()

	* demo/vw_example_4.R: Add library(pROC), add a legend to ROC plot
	* demo/00Index: Add vw_example4

	* R/plot.R (plotDensity): add utils::globalVariables() for R CMD check

	* R/dt2vw.R (dt2vw): If dependent variable is numeric, do not
	check factor levels

	* demo/vw_example_5.R: New regression example

2016-08-21  Dirk Eddelbuettel  <edd@debian.org>

	* DESCRIPTION (Suggests): Added caret (confusion matrix), ggplot2
	and earth (etitanic data)

	* R/vw.R (vw): Steps towards richer return objects, more arguments
	documented, all temp files now in current directory, new option to
	keep files
	* man/vw.Rd: Updated accordingly

	* R/plot.R: New simple density plot function
	* man/plotDensity.Rd: Documentation

	* demo/vw_example_4.R: Now plots ROC curve with (up to) three models

2016-08-20  Dirk Eddelbuettel  <edd@debian.org>

	* R/init.R (.getVW,.getPerf): Helper accessors for vw and perf
	binaries

	* R/vw.R (vw): Use vw and perf binaries stored in package
	environment, ouput simplification

2016-08-19  Dirk Eddelbuettel  <edd@debian.org>

	* R/vw.R (vw): Test for vw and perf binaries

2016-08-18  Dirk Eddelbuettel  <edd@debian.org>

	* R/init.R (.onAttach): Startup code to look for 'vw' and 'perf'

	* R/vw.R: Reindented and other whitespace changes, now uses '<-'
	assignments and TRUE/FALSE not T/F, removed @export/@import
	* R/dt2vw.R: Ditto

2016-08-17  Dirk Eddelbuettel  <edd@debian.org>

	* DESCRIPTION (License): Changed to BSD_3_clause which is what
	vowpal wabbit itself uses
	* LICENSE: Ditto

	* demo/vw_example.R: Updated (old) example, no longer uses system,
	files in R's temp directory

	* demo/vw_example_2.R: Updated (old) example_2

	* inst/examples/bostonHousing/: New (simple) regression example

2016-08-16  Dirk Eddelbuettel  <edd@debian.org>

	* DESCRIPTION (Version): 0.1.1 now passes R CMD check

	* DESCRIPTION: Rewritten / adapted to current standards
	* NAMESPACE: Updated
	* LICENSE: Added per BSD_2_clause requirements
	* R/vw.R: Roxygen documentation update, minimal changes
	* R/dt2vw.R: Ditto
	* man/vw.Rd: Updated
	* man/dt2vw.Rd: Ditto

	* DESCRIPTION (Version): 0.1.0 committed 'as is' as intial commit




src/helpers.cpp
#include "vw.h"

#include <Rcpp.h>
#include "helpers.h"

#include <fstream>
#include <stdio.h>
#include <string.h>
#include <iostream>
#include <streambuf>
#include <sstream>
#include <stdlib.h>


#include <RApiSerializeAPI.h>

#include <R.h>
#include <R_ext/Rdynload.h>

extern "C" {
#include "md5.h"
}

#ifdef _WIN32
#define PATH_SEPARATOR '\\' 
#else 
#define PATH_SEPARATOR '/' 
#endif 


// Based on code from R digest package http://dirk.eddelbuettel.com/code/digest.html
// Copyright (C) 2003 - 2016  Dirk Eddelbuettel <edd@debian.org>
std::string md5sum(char * char_x, uint32_t nChar) {
    char output[33+1];
    md5_context ctx;
    unsigned char md5sum[16];
    int j;
    md5_starts( &ctx );
    md5_update( &ctx, (uint8 *) char_x, nChar);
    md5_finish( &ctx, md5sum );
    memcpy(output, md5sum, 16);
    
    for (j = 0; j < 16; j++)
        sprintf(output + j * 2, "%02x", md5sum[j]);
    
    std::string output_str(output);
    
    return(output_str);
}

Rcpp::String check_data(Rcpp::List & vwmodel, std::string & valid_data_str, SEXP data, bool quiet, std::string mode,
                                 Rcpp::Nullable<SEXP *> namespaces, Rcpp::Nullable<Rcpp::CharacterVector> keep_space,
                                 Rcpp::Nullable<Rcpp::CharacterVector> fixed, 
                                 Rcpp::Nullable<Rcpp::CharacterVector> targets, Rcpp::Nullable<Rcpp::CharacterVector> probabilities,
                                 Rcpp::Nullable<Rcpp::String> weight, Rcpp::Nullable<Rcpp::String> base,
                                 Rcpp::Nullable<Rcpp::String> tag, Rcpp::Nullable<int> multiline) {
    
    // Check if:
    //  no previous parser options are in the model
    //  OR
    //  any of passed parser options is NOT NULL (possible when using "predict.vw" function)
    if(Rf_isNull(vwmodel["parser_opts"]) || (!Rf_isNull(namespaces) || !Rf_isNull(keep_space) || !Rf_isNull(fixed) || !Rf_isNull(targets) ||
       !Rf_isNull(probabilities) || !Rf_isNull(weight) || !Rf_isNull(base) ||
       !Rf_isNull(tag) || !Rf_isNull(multiline))) { 
       
       // In this case we want to use parser options that were passed with "vwtest" so we save parser options
       vwmodel["parser_opts"] = Rcpp::List::create(Rcpp::Named("namespaces") = namespaces , Rcpp::Named("keep_space") = keep_space,
                                                    Rcpp::Named("fixed") = fixed,
                                                    Rcpp::Named("targets") = targets, Rcpp::Named("probabilities") = probabilities,
                                                    Rcpp::Named("weight") = weight, Rcpp::Named("base") = base,
                                                    Rcpp::Named("tag") = tag, Rcpp::Named("multiline") = multiline);
       
    } else {
        // In this case we use previously saved parser options
        if(!quiet){
            Rcpp::Rcout << "Using parser options from the previous session" << std::endl;
        }
    }
    
    Rcpp::String data_md5sum("");
    uint32_t nChar;
    char * char_x;
    if(TYPEOF(data) == STRSXP) {
        // Use path to file as model input
        valid_data_str = Rcpp::as<std::string>(data);
        
        // Check path for whitespace
        if(valid_data_str.find_first_of("\t\n ") != valid_data_str.npos) {
            Rcpp::stop("Whitespace characters are not allowed in `data` path");
        }
        
        std::ifstream data_instream(valid_data_str);
        std::string data_contents((std::istreambuf_iterator<char>(data_instream)), 
                             std::istreambuf_iterator<char>());
        
        char_x = &data_contents[0u];
        nChar = data_contents.length();
        data_md5sum = md5sum(char_x, nChar);
    } else if(TYPEOF(data) == VECSXP) {
        // Parse data frame and use VW file as model input
        
        // Update valid data string
        valid_data_str = Rcpp::as<std::string>(vwmodel["dir"]) + PATH_SEPARATOR + mode + ".vw";
        // Compute md5sum of data.frame
        Rcpp::RawVector x = serializeToRaw(data);
        char_x = (char*) RAW(x);
        nChar = XLENGTH(x);
        
        data_md5sum = md5sum(char_x, nChar);
        
        // Compare new md5sum with old md5sum
        Rcpp::List vwmodel_md5sums = vwmodel["data_md5sum"];
        Rcpp::String model_md5sum = vwmodel_md5sums[mode];
        
        
        if (model_md5sum != data_md5sum) {
            if(!quiet){
                Rcpp::Rcout << "Converting data.frame to VW format" << std::endl;
            }
            Rcpp::Environment env("package:rvw");
            Rcpp::Function r_df2vw = env["df2vw"];
            // Convert data.frame to VW
            
            Rcpp::List saved_parser_opts = vwmodel["parser_opts"];
            
            // Rcpp::Nullable<SEXP *> valid_namespaces = saved_parser_opts["namespaces"];
            // Rcpp::Nullable<Rcpp::CharacterVector> valid_keep_space = saved_parser_opts["keep_space"];
            // Rcpp::Nullable<Rcpp::CharacterVector> valid_targets = saved_parser_opts["targets"];
            // Rcpp::Nullable<Rcpp::CharacterVector> valid_probabilities = saved_parser_opts["probabilities"];
            // Rcpp::Nullable<Rcpp::String> valid_weight = saved_parser_opts["weight"];
            // Rcpp::Nullable<Rcpp::String> valid_base = saved_parser_opts["base"];
            // Rcpp::Nullable<Rcpp::String> valid_tag = saved_parser_opts["tag"]; 
            // Rcpp::Nullable<int> valid_multiline = saved_parser_opts["multiline"];
            
            r_df2vw(data, valid_data_str,
                    saved_parser_opts["namespaces"], saved_parser_opts["keep_space"], saved_parser_opts["fixed"],
                    saved_parser_opts["targets"], saved_parser_opts["probabilities"],
                    saved_parser_opts["weight"], saved_parser_opts["base"], saved_parser_opts["tag"], saved_parser_opts["multiline"],
                    false
            );
        }
        
    } else {
        Rcpp::stop("Only String and data.frame types are supported");
    }
    return data_md5sum;
}

// Get number of examples used in model
int get_num_example(vw& all) {
    return all.sd->example_number + all.sd->weighted_holdout_examples;
}

bool file_exists(std::string file_name)
{
    std::ifstream infile (file_name.c_str());
    return infile.good();
}

// setup function from VW main.cc file
// modified to work in library mode using Rcpp
vw* setup_model(std::string args_str) {
    
    int argc;
    char** argv = VW::get_argv_from_string(args_str, argc);
    
    vw* all = nullptr;
    try { all = VW::initialize(argc, argv);
    }
    catch(const VW::vw_exception& ex){
        Rcpp::Rcout << ex.what() << std::endl;
        throw;
    }
    catch(...)
    {
        Rcpp::Rcout << "unknown exception" << std::endl;
        throw;
    }
    
    if (!all->quiet && !all->bfgs && !all->searchstr && !all->opts_n_args.vm.count("audit_regressor"))
    {
        Rcpp::Rcout << std::left
                    << std::setw(shared_data::col_avg_loss) << std::left << "average"
                    << " "
                    << std::setw(shared_data::col_since_last) << std::left << "since"
                    << " "
                    << std::right
                    << std::setw(shared_data::col_example_counter) << "example"
                    << " "
                    << std::setw(shared_data::col_example_weight) << "example"
                    << " "
                    << std::setw(shared_data::col_current_label) << "current"
                    << " "
                    << std::setw(shared_data::col_current_predict) << "current"
                    << " "
                    << std::setw(shared_data::col_current_features) << "current"
                    << std::endl;
        Rcpp::Rcout << std::left
                    << std::setw(shared_data::col_avg_loss) << std::left << "loss"
                    << " "
                    << std::setw(shared_data::col_since_last) << std::left << "last"
                    << " "
                    << std::right
                    << std::setw(shared_data::col_example_counter) << "counter"
                    << " "
                    << std::setw(shared_data::col_example_weight) << "weight"
                    << " "
                    << std::setw(shared_data::col_current_label) << "label"
                    << " "
                    << std::setw(shared_data::col_current_predict) << "predict"
                    << " "
                    << std::setw(shared_data::col_current_features) << "features"
                    << std::endl;
    }
    
    return all;
}	

// Collect final performance evaluation results
Rcpp::List get_eval(vw& all)
{
    int num_examples = all.sd->example_number;
    double weighted_example_sum = all.sd->weighted_examples();
    double weighted_label_sum = all.sd->weighted_labels;
    double avg_loss = NA_REAL;
    double avg_multiclass_log_loss = NA_REAL;
    float best_const = NA_REAL;
    float best_const_loss = NA_REAL;
    int total_feature = all.sd->total_features;
    
    if(all.holdout_set_off) {
        if (all.sd->weighted_labeled_examples > 0) {
            avg_loss = all.sd->sum_loss / all.sd->weighted_labeled_examples; 
        } else {
            avg_loss = NA_REAL; 
        }
    } else if((all.sd->holdout_best_loss == FLT_MAX) || (all.sd->holdout_best_loss == FLT_MAX * 0.5)) {
        avg_loss = NA_REAL;
    } else {
        avg_loss = all.sd->holdout_best_loss;
    }
    if (all.sd->report_multiclass_log_loss)
    {
        if (all.holdout_set_off) {
            avg_multiclass_log_loss = all.sd->multiclass_log_loss / all.sd->weighted_labeled_examples; 
        } else {
            avg_multiclass_log_loss = all.sd->holdout_multiclass_log_loss / all.sd->weighted_labeled_examples;
        }
    }
    // Get best_const and best_const_loss
    copy_get_best_constant(all, best_const, best_const_loss);
    
    Rcpp::List eval_list = Rcpp::List::create(
        Rcpp::Named("num_examples") = num_examples,
        Rcpp::Named("weighted_example_sum") = weighted_example_sum,
        Rcpp::Named("weighted_label_sum") = weighted_label_sum,
        Rcpp::Named("avg_loss") = avg_loss,
        Rcpp::Named("avg_multiclass_log_loss") = avg_multiclass_log_loss,
        Rcpp::Named("best_const") = best_const,
        Rcpp::Named("best_const_loss") = best_const_loss,
        Rcpp::Named("total_feature") = total_feature
        );
    
    return(eval_list);
}

// Copy of get_best_constant function from best_constant.cc file
bool copy_get_best_constant(vw& all, float& best_constant, float& best_constant_loss)
{
    if (all.sd->first_observed_label == FLT_MAX || // no non-test labels observed or function was never called
        (all.loss == nullptr) || (all.sd == nullptr)) return false;
    
    float label1 = all.sd->first_observed_label; // observed labels might be inside [sd->Min_label, sd->Max_label], so can't use Min/Max
    float label2 = (all.sd->second_observed_label == FLT_MAX)?0: all.sd->second_observed_label; // if only one label observed, second might be 0
    if (label1 > label2) {float tmp = label1; label1 = label2; label2 = tmp;} // as don't use min/max - make sure label1 < label2
    
    float label1_cnt;
    float label2_cnt;
    
    if (label1 != label2)
    {
        label1_cnt = (float) (all.sd->weighted_labels - label2*all.sd->weighted_labeled_examples)/(label1 - label2);
        label2_cnt = (float)all.sd->weighted_labeled_examples - label1_cnt;
    }
    else
        return false;
    
    if ( (label1_cnt + label2_cnt) <= 0.) return false;
    
    
    po::variables_map& vm = all.opts_n_args.vm;
    
    std::string funcName;
    if(vm.count("loss_function"))
        funcName = vm["loss_function"].as<std::string>();
    else
        funcName = "squared";
    
    if(funcName.compare("squared") == 0 || funcName.compare("Huber") == 0 || funcName.compare("classic") == 0)
        best_constant = (float) all.sd->weighted_labels / (float) (all.sd->weighted_labeled_examples);
    else if (all.sd->is_more_than_two_labels_observed)
    {
        //loss functions below don't have generic formuas for constant yet.
        return false;
        
    }
    else if(funcName.compare("hinge") == 0)
    {
        
        best_constant = label2_cnt <= label1_cnt ? -1.f: 1.f;
        
    }
    else if(funcName.compare("logistic") == 0)
    {
        
        label1 = -1.; //override {-50, 50} to get proper loss
        label2 =  1.;
        
        if (label1_cnt <= 0) best_constant = 1.;
        else if (label2_cnt <= 0) best_constant = -1.;
        else
            best_constant = std::log(label2_cnt/label1_cnt);
        
    }
    else if(funcName.compare("quantile") == 0 || funcName.compare("pinball") == 0 || funcName.compare("absolute") == 0)
    {
        
        float tau = 0.5;
        if(vm.count("quantile_tau"))
            tau = vm["quantile_tau"].as<float>();
        
        float q = tau*(label1_cnt + label2_cnt);
        if (q < label2_cnt) best_constant = label2;
        else best_constant = label1;
    }
    else
        return false;
    
    if (!all.sd->is_more_than_two_labels_observed)
    {
        best_constant_loss =  (label1_cnt>0)?all.loss->getLoss(all.sd, best_constant, label1) * label1_cnt:0.0f;
        best_constant_loss += (label2_cnt>0)?all.loss->getLoss(all.sd, best_constant, label2) * label2_cnt:0.0f;
        best_constant_loss /= label1_cnt + label2_cnt;
    }
    else best_constant_loss = FLT_MIN;
    
    return true;
}

std::vector<std::string> split_str(const std::string &s, char del) {
    std::stringstream s_stream(s);
    std::string item;
    std::vector<std::string> elems;
    while ( getline(s_stream, item, del) ) {
        elems.push_back(item);
    }
    return elems;
}
